pkg/cmd/server/kubernetes/network/transport/transport.go

```diff
@@ -0,0 +1,212 @@
+	if clientConfig.Transport != nil {
+		return fmt.Errorf("there is already a transport configured")
+	}
+	tlsConfig, err := restclient.TLSConfigFor(clientConfig)
+	if err != nil {
+		return fmt.Errorf("unable to configure TLS for the rest client: %v", err)
+	}
+	if tlsConfig == nil {
+		tlsConfig = &tls.Config{}
+	}
+	manager := &certificateManager{store: store, minimumRefresh: period}
+	tlsConfig.Certificates = nil
+	tlsConfig.GetClientCertificate = func(requestInfo *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		cert := manager.Current()
+		if cert == nil {
+			return &tls.Certificate{Certificate: nil}, nil
+		}
+		return cert, nil
+	}
+
+	// Custom dialer that will track all connections it creates.
+	t := &connTracker{
+		dialer: &net.Dialer{Timeout: 30 * time.Second, KeepAlive: 30 * time.Second},
+		conns:  make(map[*closableConn]struct{}),
+	}
+
+	lastCert := manager.Current()
+	go wait.Until(func() {
+		curr := manager.Current()
+		if curr == nil || lastCert == curr {
+			// Cert hasn't been rotated.
+			return
+		}
+		lastCert = curr
+
+		glog.Infof("certificate rotation detected, shutting down client connections to start using new credentials")
+		// The cert has been rotated. Close all existing connections to force the client
+		// to reperform its TLS handshake with new cert.
+		//
+		// See: https://github.com/kubernetes-incubator/bootkube/pull/663#issuecomment-318506493
+		t.closeAllConns()
+	}, period, stopCh)
+
+	clientConfig.Transport = utilnet.SetTransportDefaults(&http.Transport{
+		Proxy:               http.ProxyFromEnvironment,
+		TLSHandshakeTimeout: 10 * time.Second,
+		TLSClientConfig:     tlsConfig,
+		MaxIdleConnsPerHost: 25,
+		DialContext:         t.DialContext, // Use custom dialer.
+	})
+
+	// Zero out all existing TLS options since our new transport enforces them.
+	clientConfig.CertData = nil
+	clientConfig.KeyData = nil
+	clientConfig.CertFile = ""
+	clientConfig.KeyFile = ""
+	clientConfig.CAData = nil
+	clientConfig.CAFile = ""
+	clientConfig.Insecure = false
+	return nil
+}
+
+// certificateManager reloads the requested certificate from disk when requested.
+type certificateManager struct {
+	store          certificate.Store
+	minimumRefresh time.Duration
+
+	lock      sync.Mutex
+	cert      *tls.Certificate
+	lastCheck time.Time
+}
+
+// Current retrieves the latest certificate from disk if it exists, or nil if
+// no certificate could be found. The last successfully loaded certificate will be
+// returned.
+func (m *certificateManager) Current() *tls.Certificate {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// check whether the cert has expired and whether we've waited long enough since our last
+	// check to look again
+	cert := m.cert
+	if cert != nil {
+		now := time.Now()
+		if now.After(cert.Leaf.NotAfter) {
+			if now.Sub(m.lastCheck) > m.minimumRefresh {
+				glog.V(2).Infof("Current client certificate is expired, checking from disk")
+				cert = nil
```

Comment:
After you've set this, you're setting yourself up to hotloop on this aren't you?  I'm not seeing another check against `lastCheck` after the cert is nil.

Reply:
Sets function cert, not m.cert.  We always return the last cert we loaded.